// observer/node_modules/checkSquares.js
this.Board;

function checker(board) {

}

function checkSquares(square, moveFile, moveRank) {
  var name = square.getPieceName();
  switch(name){
    case 'P': return pawnSquares(square,moveFile,moveRank);
    case 'S': return knightSquares(square,moveFile,moveRank);
    case 'B': return bishopSquares(square,moveFile,moveRank);
    case 'R': return rookSquares(square,moveFile,moveRank);
    case 'Q': return queenSquares(square,moveFile,moveRank);
    case 'K': return queenSquares(square,moveFile,moveRank);
    default : return false;
  }

}

function pawnSquares(square,moveFile,moveRank){
    if(square.getFile() == moveFile
       && this.Board.getPiece(moveFile,moveRank) != '')
       return false;
}

function knightSquares(square,moveFile,moveRank){
    if(this.Board.getPiece(moveFile,moveRank).getColor()
        == square.getPiece().getColor())
        return false;
    return true;
}

function bishopSquares(square,moveFile,moveRank){
    var file = square.getFile();
    var rank = square.getRank();

    if(moveFile > file && moveRank > rank)
        for(var i =file , j = rank; i < moveFile, j < rank; i++, j++)
            if(this.Board.getPiece(i,j) != '' )
                return false;

    if(moveFile < file && moveRank < rank)
        for(var i =file , j = rank; i > moveFile, j > rank; i--, j--)
            if(this.Board.getPiece(i,j) != '' )
                return false;

    if(moveFile > file && moveRank < rank)
        for(var i =file , j = rank; i < moveFile, j > rank; i++, j--)
            if(this.Board.getPiece(i,j) != '' )
                return false;

    if(moveFile < file && moveRank > rank)
        for(var i =file , j = rank; i > moveFile, j < rank; i--, j++)
            if(this.Board.getPiece(i,j) != '' )
                return false;

    if(this.Board.getPiece(moveFile,moveRank).getColor()
        == square.getPiece().getColor())
        return false();
    return true;

}

function rookSquares(square,moveFile,moveRank){
    var file = square.getFile();
    var rank = square.getRank();

    if(file == moveFile){
        if(rank>moveRank)
            for(var i = moveRank; i < rank; i++)
                if(this.Board.getPiece(moveFile,i) != '')
                    return false;
        else
            for( i = rank; i< moveRank; i++)
                if(this.Board.getPiece(moveFile,i) != '')
                    return false;
    }

    if(rank == moveRank){
        if(file>moveFile)
            for( i = moveFile; i < file; i++)
                if(this.Board.getPiece(i.moveRank) != '')
                    return false;
        else
            for( i = file; i< moveFile; i++)
                if(this.Board.getPiece(i,moveRank) != '')
                    return false;
    }
    if(this.Board.getSquare(moveFile,moveRank).piece().getColor()
        == square.getPiece().getColor())
        return false();

    return true;

}

function queenSquares(square,moveFile,moveRank){

    var file = square.getFile();
    var rank = square.getRank();

    if(moveFile > file && moveRank > rank)
        for(var i =file , j = rank; i < moveFile, j < rank; i++, j++)
            if(this.Board.getPiece(i,j) != '' )
                return false;

    if(moveFile < file && moveRank < rank)
        for(var i =file , j = rank; i > moveFile, j > rank; i--, j--)
            if(this.Board.getPiece(i,j) != '' )
                return false;

    if(moveFile > file && moveRank < rank)
        for(var i =file , j = rank; i < moveFile, j > rank; i++, j--)
            if(this.Board.getPiece(i,j) != '' )
                return false;

    if(moveFile < file && moveRank > rank)
        for(var i =file , j = rank; i > moveFile, j < rank; i--, j++)
            if(this.Board.getPiece(i,j) != '' )
                return false;

     if(file == moveFile){
        if(rank>moveRank)
            for(var i = moveRank; i < rank; i++)
                if(this.Board.getPiece(moveFile,i) != '')
                    return false;
        else
            for( i = rank; i< moveRank; i++)
                if(this.Board.getPiece(moveFile,i) != '')
                    return false;
    }

    if(rank == moveRank){
        if(file>moveFile)
            for( i = moveFile; i < file; i++)
                if(this.Board.getPiece(i.moveRank) != '')
                    return false;
        else
            for( i = file; i< moveFile; i++)
                if(this.Board.getPiece(i,moveRank) != '')
                    return false;
    }
    if(this.Board.getPiece(moveFile,moveRank).getColor()
        == square.getPiece().getColor())
        return false();

    return true;
}

module.exports = checkSquares;
