// observer/node_modules/legalChecker.js
var king = require('../../pieces');
var rook = require('../../pieces');
var checkMate = require('./checkMate');
var checkSquares = require('./checkSquares');

function isLegal(square, moveFile, moveRank, observer) {
    // name of the piece
    var name = square.getPieceName();
    var color = square.getPiece().getColor();

    //the boolean for moving(true/false)sssss
    var bool;

    //go method given by the piece we have
    switch(name) {
        case 'P':  bool = movePawn(square ,moveFile, moveRank,observer,color);
                   break;
        case 'S':  bool = moveKnight(square,moveFile, moveRank,observer,color);
                   break;
        case 'B':  bool = moveBishop(square,moveFile,moveRank,observer,color);
                   break;
        case 'R':  bool = moveRook(square,moveFile,moveRank,observer,color);
                   break;
        case 'Q':  bool = moveQueen(square,moveFile,moveRank,observer,color);
                   break;
        default : return false;
    }


    // if(bool == true && king.isAttacked)
    // {
        // king.isChecked();
    // }

//    if(checkMate(king,color) && king.isChecked == true )
//            return "CheckMate";

    return bool;
}

function movePawn(square,moveFile, moveRank, observer,color) {

     if(color == 1)
        if(observer.pieces.getKingWhite().isAttacked())
          return false;
     else
        if(observer.pieces.getKingBlack().isAttacked())
          return false;

     if(checkSquares(square,moveFile,moveRank,observer))
        return false;

    // if(king.isChecked)
        //return false;


    if (square.getPiece().isFirstMove()
        && ( Math.abs(moveRank - square.rank()) == 2
             && moveFile - conv(square.file()) == 0) ) {
        observer.board.move(square.getPiece(),moveFile,moveRank);
        return true;
    }

    if (Math.abs(moveRank-square.rank()) == 1
        && moveFile - conv(square.file()) == 0) {
        observer.board.move(square.getPiece(),moveFile,moveRank);
        return true;
    }

    return false;
}

function moveKnight(square,moveFile, moveRank,observer){

    // if(king.isAttacked())
        // return false;

    if(!checkSquares(square,moveFile,moveRank, observer))
        return false;

    // if(king.isChecked)
        // return false;

    if (Math.abs(moveFile)  == 1 && Math.abs(moveRank) == 2 ||
        Math.abs(moveFile)  == 2 && Math.abs(moveRank) == 1){
       observer.board.move(square.getPiece(), moveFile,moveRank);
       return true;}
    return false;
}

function moveBishop(square,moveFile, moveRank,observer){

    // if(king.isAttacked())
        // return false;

    if(!checkSquares(square,moveFile,moveRank,observer))
        return false;

    // if(king.isChecked)
        // return false;

    if(Math.abs(moveFile-square.getFile())
        == Math.abs(moveRank-square.getRank())){
        observer.board.move(square.getPiece(), moveFile,moveRank);
        return true; }
    return false;
}

function moveRook(square,moveFile, moveRank,observer){

    // if(king.isAttacked())
        // return false;

    if(!checkSquares(square,moveFile,moveRank))
        return false;

    // if(king.isChecked)
        // return false;

    if(moveFile == square.getFile || moveRank == square.getRank)
    {
        observer.board.move(square.getPiece(), moveFile,moveRank);

        square.getPiece().moved();
        return true;}

    return false;
}

function moveQueen(square, moveFile, moveRank,observer){

    // if(king.isAttacked())
        // return false;

    if(!checkSquares(square,moveFile,moveRank))
        return false;

    // if(king.isChecked)
        // return false;

    if((moveFile == square.getFile || moveRank == square.getRank)
        && (Math.abs(moveFile-square.getFile())
        == Math.abs(moveRank-square.getRank()))){

        observer.board.move(square.getPiece(), moveFile,moveRank);
        return true; }
    return false;

}

function moveKing(square,moveFile,moveRank,observer){


    if(!checkSquares(square,moveFile,moveRank))
        return false;


    if(square.getPiece().getColor == 1)
    {
        if(moveFile == 'g' && moveRank == '1' && !square.getPiece.isMoved()
            && !observer.board.getPiece('h',1).isMoved()
            &&  checkSquares(observer.board.getPiece('g',1),'e',1)
            && !observer.board.getPiece('g',1).isAttacked()
            && !observer.board.getPiece('f',1).isAttacked()){

            observer.board.move('K', 'g', 1 );
            observer.board.move('R', 'f', 1 );
            observer.board.getPiece().moved();
            observer.board.getPiece('f',1).moved();
            return true; }

        else
        if(moveFile == 'c' && moveRank == '1' && !square.getPiece.isMove()
            && !observer.board.getPiece('a',1).isMoved()
            &&  checkSquares(observer.board.getPiece('a',1),'e',1)
            && !observer.board.getPiece('b',1).isAttacked()
            && !observer.board.getPiece('c',1).isAttacked()
            && !observer.board.getPiece('d',1).isAttacked()){

            observer.board.move('K','c' ,1);
            observer.board.move('R','d', 1);
            square.getPiece().moved();
            observer.board.getPiece('d',1).moved();
            return true; }

    }
    else
    {
        if(moveFile == 'g' && moveRank == '8' && !square.getPiece().isMoved()
            && !observer.board.getPiece('h',8).isMoved()
            &&  checkSquares(observer.board.getPiece('g',8),'e',8)
            && !observer.board.getPiece('g',8).isAttacked()
            && !observer.board.getPiece('f',8).isAttacked()){

            observer.board.move('K', 'g', 8 );
            observer.board.move('R', 'f', 8 );
            observer.board.getPiece().moved();
            observer.board.getPiece('h',8).moved();
            return true; }

        else
        if(moveFile == 'c' && moveRank == '8' && !square.getPiece().isMove()
            && !observer.board.getPiece('a',8).isMoved()
            &&  checkSquares(observer.board.getPiece('a',8),'e',8)
            && !observer.board.getPiece('b',8).isAttacked()
            && !observer.board.getPiece('c',8).isAttacked()
            && !observer.board.getPiece('d',8).isAttacked()){

            observer.board.move('K','c' ,8);
            observer.board.move('R','d', 8);
            square.getPiece().moved();
            observer.board.getPiece('d',8).moved();
            return true; }

    }



    if(Math.abs(moveFile - square.getFile()) <= 1
                && Math.abs(moveRank - square.getRank()) <=1 ){
        observer.Board.move(square.getPiece(),moveFile,moveRank);
        square.getPiece().moved();
        king.isUnchecked();
        return true; }
    return false;
}

function conv(file)
{
    switch(file) {
        case 'a': return 1;
        case 'b': return 2;
        case 'c': return 3;
        case 'd': return 4;
        case 'e': return 5;
        case 'f': return 6;
        case 'g': return 7;
        case 'h': return 8;
        default : return 0; }
}

module.exports = isLegal;
